File: slf4j-messager/src/main/java/io/openapiprocessor/slf4j/messager/MessagerLoggerConfiguration.java
/*
 * Copyright 2022 https://github.com/openapi-processor/openapi-processor-base
 * PDX-License-Identifier: Apache-2.0
 */

package io.openapiprocessor.slf4j.messager;

import org.slf4j.event.Level;

import java.io.IOException;
import java.util.Properties;

public class MessagerLoggerConfiguration {
    private static final String CONFIG_FILE_NAME = "messagerlogger.properties";

    private boolean initialized = false;

    private final Properties properties = new Properties();

    private Level logLevel = Level.INFO;


    void init() {
        if (initialized) {
            return;
        }

        loadProperties ();
        initLogLevel ();
        initialized = true;
    }

    Level getLogLevel() {
        return logLevel;
    }

    private void initLogLevel () {
        String ll = properties.getProperty ("log.level");
        if (ll == null) {
            return;
        }

        switch (ll.toUpperCase ()) {
            case "ERROR":
                logLevel = Level.ERROR;
                break;
            case "WARN":
                logLevel = Level.WARN;
                break;
            case "INFO":
                logLevel = Level.INFO;
                break;
            case "DEBUG":
                logLevel = Level.DEBUG;
                break;
            case "TRACE":
                logLevel = Level.TRACE;
                break;
        }
    }

    private void loadProperties () {
        try {
            ClassLoader classLoader = Thread.currentThread ().getContextClassLoader ();
            properties.load (classLoader.getResourceAsStream (CONFIG_FILE_NAME));
        } catch (IOException ignored) {
            // ignore, use defaults
        }
    }
}

File: openapi-processor-core/src/main/kotlin/io/openapiprocessor/core/logging/MessagerLoggerConfiguration.kt
/*
 * Copyright 2022 https://github.com/openapi-processor/openapi-processor-base
 * PDX-License-Identifier: Apache-2.0
 */
package io.openapiprocessor.core.logging

import org.slf4j.event.Level
import java.io.IOException
import java.util.*

/**
 * log leve can be configured with `messagerlogger.properties`. It does allow a single property:
 * ```
 * log.level=error|warn|info|debug|trace
 * ```
 */
class MessagerLoggerConfiguration {
    private var initialized = false
    private val properties = Properties()

    var logLevel = Level.INFO

    fun init() {
        if (initialized) {
            return
        }
        loadProperties()
        initLogLevel()
        initialized = true
    }

    private fun initLogLevel() {
        val ll = properties.getProperty("log.level") ?: return

        when (ll.uppercase(Locale.getDefault())) {
            "ERROR" -> logLevel = Level.ERROR
            "WARN" -> logLevel = Level.WARN
            "INFO" -> logLevel = Level.INFO
            "DEBUG" -> logLevel = Level.DEBUG
            "TRACE" -> logLevel = Level.TRACE
        }
    }

    private fun loadProperties() {
        try {
            val classLoader = Thread.currentThread().contextClassLoader
            properties.load(classLoader.getResourceAsStream(CONFIG_FILE_NAME))
        } catch (ignored: IOException) {
            // ignore, use defaults
        }
    }

    companion object {
        private const val CONFIG_FILE_NAME = "messagerlogger.properties"
    }
}
